### PR TITLE
Support custom root directory for temporary directories and files

### PR DESCRIPTION
**Pull request by jambayk**
## Describe your changes
Olive uses a lot of temporary directories which are created in the default tempory directory root of the system. However, we encounter issues with storage, permissions, files disappearing, etc.
This PR adds a new olive module called `olive.tmp_dir` which allows setting/user an environment variable `OLIVE_TMP_DIR_ROOT` that will be used as the root directory for all olive based temporary directories. 

Code that set their own `dir` are unchanged. Most of them use the destination dir as the `dir` since that allows for fast move through rename (shutil.move under the same file system is a rename). 

## Checklist before requesting a review
- [x] Add unit tests for this change.
- [x] Make sure all tests can pass.
- [x] Update documents if necessary.
- [ ] Format your code by running `pre-commit run --all-files`
- [x] Is this a user-facing change? If yes, give a description of this change to be included in the release notes.

Allow user to set custom root directory for olive temporary directories and files using environment variable `OLIVE_TMP_DIR_ROOT` or workflow config option `tmp_dir_root`.

### CODE REVIEW

File: olive/workflows/run/run.py [hunk 1]
```diff
@@ -144,6 +145,10 @@ def run(config: Union[str, Path, dict], setup: bool = False, data_root: str = No
     set_default_logger_severity(config.engine.log_severity_level)
     ort.set_default_logger_severity(config.engine.ort_log_severity_level)
 
+    # set tmp dir root
+    if config.engine.tmp_dir_root is not None:
+        set_tmp_dir_root(config.engine.tmp_dir_root)
```
Comment (guotuofeng) [inline olive/workflows/run/run.py:150]: it seems we can only change the tempfile.tempdir like https://docs.python.org/3/library/tempfile.html#tempfile.tempdir? so there is no need tmp_dir.py?

**Review by guotuofeng — CHANGES_REQUESTED** [review]
https://docs.python.org/3/library/tempfile.html#tempfile.tempdir should be able to meet our requirements.

File: olive/workflows/run/config.py [hunk 2]
```diff
@@ -34,6 +34,7 @@ class RunEngineConfig(EngineConfig):
     packaging_config: PackagingConfig = None
     log_severity_level: int = 1
     ort_log_severity_level: int = 3
+    tmp_dir_root: str = None  # if None, will use the whatever is set in OLIVE_TMP_DIR_ROOT
```
Comment (guotuofeng) [inline olive/workflows/run/config.py:37]: should this belong engine config or run config? BTW, need the parameter passed by command line like:
python -m olive.workflows.run --config <config_path> --temp_dir blalblabla?

**Comment by jambayk** [comment]
closing in favor of https://github.com/microsoft/Olive/pull/683

**Pull request closed without merging**